### PR TITLE
[browser-wallet] Add UI for Open Channel Objective

### PR DESCRIPTION
**Pull request by kerzhner**
Browser wallet UI can be:
- xstate machine based. This is the established UI methodology in the browser wallet.
- objective based. This is added in this PR.

For objective based UI, the following are the ways that the `ChannelWallet` communicates with the UI and the UI communicates with the `ChannelWallet`:
- On initialization, the `ChannelWallet` is supplied an `updateUI` function that the `ChannelWallet` must invoke whenever the `ChannelWallet` needs the UI to update.
- The UI is supplied a function that can be invoked with an Objective event argument. This function is wired up on `ChannelWallet` initialization.

This PR has been tested with:
- existing unit and interop tests.
- storybook to display new UI.
- manual calls to browser wallet APIs.

https://user-images.githubusercontent.com/1062379/115122417-1ce9b380-9f75-11eb-84bc-9510dd34082a.mov

Fixes https://github.com/statechannels/statechannels/issues/3497.

### CODE REVIEW

File: packages/browser-wallet/src/channel-wallet-types.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import {DirectFunder, SharedObjective, SignedState} from '@statechannels/wallet-core';
+import {Interpreter} from 'xstate';
+
+export type AnyInterpreter = Interpreter<any, any, any>;
+export type OnObjectiveEvent = (event: DirectFunder.OpenChannelEvent) => void;
```
Comment (lalexgap) [inline packages/browser-wallet/src/channel-wallet-types.ts:5]: The name`OnObjectiveEvent` seems to indicate an event handler but it's actually used by the UI to trigger an objective event. Could we name it something like `triggerObjectiveEvent` ?

File: packages/browser-wallet/src/channel-wallet-types.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * The channel wallet is not able to update UI. The channel wallet is supplied a callback to invoke
+ * when UI should be updated.
+ *
+ * UI needs a way to communicate objective events to the channel wallet. UI invokes onObjectiveEvent
+ * when UI triggers an objetive event
```
Comment (lalexgap) [inline packages/browser-wallet/src/channel-wallet-types.ts:11]:
```suggestion
 * when UI triggers an objective event
```

File: packages/browser-wallet/src/channel-wallet.ts [hunk 3]
```diff
@@ -32,35 +31,38 @@ import {ADD_LOGS, WALLET_VERSION} from './config';
...
-  objectives: SharedObjective[];
-  signedStates: SignedState[];
-};
-
 export class ChannelWallet {
   public workflows: Workflow[];
+  /**
+   *
+   * @param chainAddress Ethereum address for the wallet
+   * @param updateUI Callback that will be invoked when the channel wallet wants to update UI
```
Comment (lalexgap) [inline packages/browser-wallet/src/channel-wallet.ts:39]: Could the docs be updated to describe the `setOnObjectiveEvent` param?

**Review by lalexgap — APPROVED** [review]

File: packages/browser-wallet/src/channel-wallet-types.ts (hunk 1 above)
Comment (kerzhner) [reply packages/browser-wallet/src/channel-wallet-types.ts]: Good call. Addressed in 9141a23

**Pull request merged**